packages/nimble-components/package.json

```diff
@@ -74,6 +74,8 @@
     "d3-scale": "^4.0.2",
     "d3-selection": "^3.0.0",
     "d3-zoom": "^3.0.0",
```

Comment:
Moving [this comment](https://github.com/ni/nimble/pull/1335#issuecomment-1614835881) to a file thread. FYI @vivinkrishna-ni we prefer to comment on files rather than on the PR Conversation tab because it makes it easier to keep the responses in a thread

---

@jattasNI, When we added [`@tiptap/pm`](https://www.npmjs.com/package/@tiptap/pm), the `npm audit` fails, and here is the [failed build](https://github.com/ni/nimble/actions/runs/5421593611/jobs/9857224063#step:9:43) of the commit. 

So we tried running `npm audit fix` locally to fix the vulnerabilities and `npm install` to make sure the packages were installed properly but I  faced an issue with the `@storybook/addon-actions` and `@storybook/html` as in the below screenshot and `npm build` fails due to this change.
![image](https://github.com/ni/nimble/assets/123377523/64c4f3bc-8e3d-4086-bd3c-7663b2c6a189)

I think `npm audit fix` updated the version of storybook from 7.0.4 to 7.0.24 so the patch conflicts as mentioned in the image. 

As `@tiptap/pm` provides all prosemirror libraries but we require only prosemirror-markdown and prosemirror-model. So for now we eliminated this issue by installing only the required prosemirror libraries instead of all `@tiptap/pm` that causes the vulnerability  (to be specific, a vulnerability in [prosemirror-trailing-node](https://github.com/ni/nimble/actions/runs/5421593611/jobs/9857224063#step:9:43)). 

1. Can we replace the `tiptap/pm` with the original `prosemirror` libraries as mentioned above to eliminate the audit issue?
2. What might be the better approach to fix audit issues like this here? It would be really helpful if you could point us to some references to fix issues related to this.
3. Also in the current build, `npm run test` fails in Blazor components. Here is the [link](https://github.com/ni/nimble/actions/runs/5423164727/jobs/9862095300?pr=1335#step:17:6569). Is this related to the change done in this PR?

Please let me know your thoughts.

Reply:
Depending on the targeted set of packages sounds like a good fix to me for two reasons:
1. it doesn't pull in a dependency with a known vulnerability
2. (if I'm understanding correctly) it reduces the number of packages we depend on overall (more direct dependencies but fewer transitive dependencies) which is good for install time and vulnerability surface area

If you think we'll need `tiptap/pm` or `prosemirror-trailing-node` in the future, it would be good to ensure they have a bug tracking the work to adopt the fix for the vulnerability. That way hopefully they fix it (or accept a fix from us) before we pull it in. If you file a bug it's good to link to the [vulnerability report](https://github.com/advisories/GHSA-c2qf-rxjj-qqgw) from the failed pipeline.

Our [docs for dealing with audit failures](https://github.com/ni/nimble/blob/main/CONTRIBUTING.md#npm-audit) are in CONTRIBUTING. I'd say they're a bit out of date: `npm audit fix` has rarely been useful to us since it updates all our dependencies, not just the one with a known vulnerability. This is often too aggressive and causes other build issues like the ones you saw with storybook. 

For the sub-dependency vulnerability case described in the docs, we've often ended up temporarily disabling `audit` while we wait for a fix in the direct dependency.

Reply:
Thanks for your input. We will only have the necessary packages that are required for this viewer component and remove the vulnerabilities for now.

But as per the [vulnerability report](https://github.com/advisories/GHSA-c2qf-rxjj-qqgw), it seems the package which is affected, is semver v7.5.2, due to [some addition of new Range functionality](https://github.com/npm/node-semver/commit/717534ee353682f3bcf33e60a8af4292626d4441) as per the description in the report. So I have tried installing the `@tiptap/pm` package locally outside the nimble repo and tried `npm audit` which shows no vulnerabilities.

Also, I created a new branch for `nimble-rich-text-editor` and added the `@tiptap/core` and `@tiptap/starter-kit`(required packages for editor component) and tried running `npm audit` and this fails due to the same issue and [here is the failed pipeline](https://github.com/ni/nimble/actions/runs/5475978059/jobs/9972838414#step:9:16). This again gives the [same vulnerability report](https://github.com/advisories/GHSA-c2qf-rxjj-qqgw) as we encountered before. Again, locally(outside nimble repo) when I tried checking the audit for these tiptap packages, there are no vulnerabilities. 

So I am thinking this is not due to the newly added packages and some updates being added to other existing packages which are vulnerable due to the new changes like the one in semver. I am yet to see this issue in deeper but it would be really helpful to get your thoughts on this and understand if there is something I am missing here.

Reply:
The npm package version resolution algorithm is tricky. It tries to find the best version that satisfies all the requirements specified by all the package.json files in the tree, so the version that gets resolved will change depending on what other packages are requesting. 

It's likely that something else was already depending on the vulnerable version of `semver` as a dev dependency. We didn't catch this because we currently only audit prod dependencies. Then your change promoted it to a non-dev (prod) dependency at the same version as it was before. That version might be older than what gets installed if you depend on `tiptap/core` in isolation while still being compatible with the version that tiptap/core requests. But now that it's a prod dependency, npm audit catches it.

Fixing this might require finding and updating the other package that depends on an older version of `semver`. (`npm ls semver` is a helpful command for this). But I suspect we'll find that it's a dependency which is difficult to update like Angular or Storybook.

So my inclination is not to try to fix this properly right now if your current approach is passing the build and you don't see us needing more packages than what you currently depend on.

Reply:
Yes, for `nimble-rich-text-viewer` we don't need more packages and it is not a blocker for this PR as well but for `nimble-rich-text-editor` we need `@tiptap/core` as it is a required package and there it is failing in the pipeline. If it is fine to take this discussion in the rich-text-editor PR, we can do that. But I would like to share our findings about what is causing this vulnerability and a possible workaround for this issue.

The straight-away workaround we could find is to override the `semver` version to `7.5.3` as given in the below issue.
https://github.com/babel/babel/issues/15720#issuecomment-1608104971

The only difference in the `semver` package before and after installing `@tiptap/core` is that the tiptap has the dependency on the updated version (7.22.6) of `@babel/helper-compilation-targets`. In this new version, Babel has removed the dependency on semver by [creating a fork and backporting it to version 6](https://github.com/babel/babel/pull/15742/files#diff-2815a6ad8289ff68eecc05d71783c5ceb821088aae59b916f9a6a8fb04b6c015). This change in the dependency tree of semver before and after installing @tiptap/core is likely the cause of the vulnerability in the prod.

Before installing `tiptap/core`:
```
├─┬ @ni/nimble-components@19.4.4 -> .\packages\nimble-components
│ ├─┬ @babel/cli@7.20.7
│ │ └─┬ make-dir@2.1.0
│ │   └── semver@5.7.1
│ ├─┬ @babel/core@7.20.12
│ │ ├─┬ @babel/helper-compilation-targets@7.21.4
│ │ │ └── semver@6.3.0
│ │ └── semver@6.3.0
```

After installing `tiptap/core`:
```
├─┬ @ni/nimble-components@19.4.4 -> .\packages\nimble-components
│ ├─┬ @babel/cli@7.20.7
│ │ └─┬ make-dir@2.1.0
│ │   └── semver@5.7.1
│ ├─┬ @babel/core@7.20.12
│ │ └── semver@6.3.0
```

Once we begin working with the `nimble-rich-text-editor`, we will investigate this issue and find out whether have to update other packages in a safer way if it is possible. However, at present, we have the above-mentioned workaround in order to resolve the pipeline blockage, and here is the [success pipeline](https://github.com/ni/nimble/actions/runs/5485230212/jobs/9993790703#step:9:12) to it.

Reply:
I think deferring this until the editor PR is a good approach.

I've seen the `overrides` feature before but never got it to work myself. But if it is working then it sounds like a good option: a targeted way to update a sub-dependency without waiting for fixes from the direct dependency. We could possibly even leverage this for other `audit` failures that we've been suppressing.

Reply:
Didn't realize overrides was added to npm in v8. Does look like it has some major issues though :/ https://github.com/npm/cli/issues/5850

Reply:
Sure, thanks for your input. We will try looking into some other better options before bringing up the `editor` component.